### PR TITLE
Tooltip split

### PR DESCRIPTION
**Pull request by Conengmo**
Split Tooltip class in two: one for simple text strings and one for more specific GeoJson/TopoJson behavior.

The second class is now called `GeoJsonTooltip`, but that's debatable. It inherits from `Tooltip`. Validation of the coupling with the geojson data is done in the `render` method.

### CODE REVIEW

File: folium/features.py [hunk 1]
```diff
@@ -602,25 +603,123 @@ def get_bounds(self):
...
+        self.fields = fields
+        self.aliases = aliases
+        self.labels = labels
+        self.toLocaleString = toLocaleString
+        if style:
+            assert isinstance(style, str), \
+                "Pass a valid inline HTML style property string to style."
+            # noqa outside of type checking.
+            self.style = style
+
```
Comment (jtbaker) [inline folium/features.py]: Would it be good to have an intermediary step from `**kwargs` to `options` as in `Tooltip`, to do some parameter checks using `parse_kwargs` to ensure valid parameters are being passed? 

For example, `sticky` is being passed `True` right now but has no effect because it is not part of the `kwargs` dict.

On leaflet's Tooltip [options](https://leafletjs.com/reference-1.3.2.html#tooltip), everything should be good to be passed except for `permanent` - this GeoJson tooltip is a different animal because it's a function that evaluates on a mouseover and it's not being directly bound to every feature.

File: folium/map.py [hunk 2]
```diff
@@ -348,144 +350,69 @@ def render(self, **kwargs):
...
+    @staticmethod
+    def parse_kwargs(kwargs):
+        """Validate the provided kwargs and return options as json string."""
+        valid_kwargs = {"pane": (str,),
+                        "offset": (tuple,),
+                        "direction": (str,),
+                        "permanent": (bool,),
+                        "sticky": (bool,),
+                        "interactive": (bool,),
+                        "opacity": (float, int)}
```
Comment (jtbaker) [inline folium/map.py]: I've noticed a few additional arguments available in the Leaflet's Tooltip [options](https://leafletjs.com/reference-1.3.2.html#tooltip) that it inherits from `DivOverlay` and `Layer` - `className` and `attribution` that I think would be worthwhile to expose in `**kwargs`.

- `attribution` will pass a string to append to the attribution corner of the window when the tooltip is displayed.
- `className` passes a string to give a class to the Tooltip that could be further used via in an inline `<style></style>` tag in the `Map` template down the road.

File: folium/features.py (hunk 1 above)
Comment (Conengmo) [reply folium/features.py]: Since in `GeoJsonTooltip` the kwargs are passed to regular `Tooltip`, they are validated in there. 

In `Tooltip`, `sticky` is added to kwargs before validating:
`kwargs.update({'sticky': sticky})`

Are there options that work with `Tooltip` but not with `GeoJsonTooltip`? It seems `permanent` indeed does nothing. We can assert that option is not entered in kwargs.

File: folium/map.py (hunk 2 above)
Comment (Conengmo) [reply folium/map.py]: If we're validating the options, then indeed we should have all available options in there. 

* I moved the dict with valid options to the class attributes since it doesn't change between instances. 
* I also added a function that converts the options to lowerCamelCase. So users can provide the options as Python style variable names and it will still work.

**Comment by Conengmo** [comment]
What do you think about the class name, `GeoJsonTooltip`? I couldn't think of something better. The class also works for `TopoJson` so maybe we can think of a more general name?

Can you think of another name for `toLocaleString`? The name should be in Python style name formatting, but `to_locale_string` is a bit convoluted IMO. How about `localize`? Users can read in the docstring what it does and that it uses Javascript's 'toLocaleString' method.

File: folium/features.py (hunk 1 above)
Comment (jtbaker) [reply folium/features.py]: @Conengmo `sticky`, or any non-named `kwargs` for that matter, don't seem to be adding to tooltip in `GeoJsonTooltip`: [nbviewer link](http://nbviewer.jupyter.org/gist/jtbaker/6979f6d669a658848bc7eccf44c3af0d)

lines [680-682](https://github.com/Conengmo/folium/blob/tooltip-split/folium/features.py#L680-L682) don't seem to be having an effect. I don't totally understand `super`, but it seems like it's allowing us to escape out to the parent Tooltip class for argument inheritance. 

Don't we still need to pass `self` to the `Tooltip.__init__()` to assign to our GeoJsonConstructor?

Comment (jtbaker) [reply folium/features.py]: Figured it out, [line 680-682](https://github.com/Conengmo/folium/blob/tooltip-split/folium/features.py#L680-L682):

`        super(GeoJsonTooltip, self).__init__(
            text='', style=style, sticky=sticky, **kwargs
        )`

needs to switch to:

`        super(GeoJsonTooltip, self).__init__(
            self, style=style, sticky=sticky, **kwargs
        )`

Swapping those out and behavior worked as expected for me. Do you think this will have any ancillary effects?

**Comment by jtbaker** [comment]
I think `GeoJsonTooltip` is the best choice - it is descriptive, and GeoJson is probably the highest number of use cases. Users can figure out that it should be used in the `folium.TopoJson` docstring.

Cool, I'll switch `toLocaleString` to `localize`. I couldn't find great documentation on JS `String.toLocaleString()` methods, but included it on the labels section in hopes that there might be some kind of browser locale translation service eventually. Currently it only really seems useful for `Number` fields.

File: folium/features.py (hunk 1 above)
Comment (Conengmo) [reply folium/features.py]: I made a mistake, sorry about that. I forgot to rename the kwargs variable in one place so the options didn't get through. Fixed that now. Now the options in `GeoJsonTooltip` get parsed by its parent.

You're right about what `super` does. You don't have to pass `self`, it does that for you. And in Python 3 you can just call `super().__init__()`, even easier.

Comment (jtbaker) [reply folium/features.py]: @Conengmo just discovered that, about `super`, and found the `self.options` issue in the `GeoJsonTooltip` as well. I made a new branch called `tooltip-split` as well, but github doesn't seem to let me PR to your `tooltip-split`.

Incorporated a few new features including a `handleOptions` function to parse non-primitive values in GeoJson properties, and renamed `toLocaleString` to `localize`.

Comment (jtbaker) [reply folium/features.py]: I will PR my `tooltip-split` into `patch-2`. It should have your commit history as well.

Comment (jtbaker) [reply folium/features.py]: I made a new branch called `tooltip-split` as well, but Github doesn't seem to let me PR to your `tooltip-split`.

Incorporated a few new features including a `handleObject` function to parse non-primitive values in GeoJson properties, and renamed `toLocaleString` to `localize`.

I will PR my tooltip-split into patch-2. It should have your commit history as well.

**Comment by jtbaker** [comment]
I made a new branch called tooltip-split as well, but Github doesn't seem to let me PR to your tooltip-split.

Incorporated a few new features including a handleObject function to parse non-primitive values in GeoJson properties, and renamed toLocaleString to localize.

I will PR my tooltip-split into patch-2. It should have your commit history as well.

**Comment by Conengmo** [comment]
Fine by me. I spotted some xml files in your PR, watch out those don't get committed.

File: folium/features.py (hunk 1 above)
Comment (Conengmo) [reply folium/features.py]: So we fixed the same bug simultaneously! I didn't stand a chance.

**Comment by jtbaker** [comment]
A question about the `camelize` utility - we're implementing here. Currently there's no documentation for any of these  keyword arguments that will be being parsed, only. If we're going to insist that the user's use `python_style_syntax` instead of `lowerCamelCase` in their arguments, I think we need to hardcode the expected values into the docstring instead of leaning on the `Leaflet` documentation.

Thoughts?

**Comment by Conengmo** [comment]
My opinion on this right now is to use Python style names all over folium, except in templates (duh) and in creating an options dictionary to be json.dumped. But users don't see that last part.

About arguments and docstrings: I don't think we should copy the entire Leaflet documentation. I think folium is a bit more high level than Leaflet and abstracts away some of the work Leaflet and its plugins do. We can include the most important options as explicit arguments, list them in the docstrings and explain how they work. Novice users can learn to use folium with the examples only, they don't have to read any Javascript and can keep using Python style notation. Power users can unlock the full power of Leaflet by reading the Leaflet documentation and providing additional kwargs. They can stick with Python style, but if they want to provide their arguments in lowerCamelCase, that's okay. If your using undocumented features you can probably manage by yourself.

So in short, folium should be simple and Python-esque for novice users, and flexible and hackable for power users.

Does that make sense to you? I'm curious where you stand on this. You seem to want to make things more rigid, and that does help in better documentation and preventing silent or ambiguous errors. But the trade-off is that things become less flexible/hackable and need more maintenance.

**Comment by jtbaker** [comment]
@Conengmo I think that makes sense. On an aspiring developer and best practices level, part of me wants there to be **one** clear, intentional way of doing things - providing multiple possible arguments that will do the same thing is confusing - which one *should* I use?

In my mind, since these values are being directly transpiled over to their Leaflet counterparts, the most clear way to do that is using the LeafletJS arguments verbatim. However, that is probably because as a part of contributing to this library, I've learned a lot more about how Leaflet and JavaScript in general, as well as Python work. As a newbie Python user I may not have understood folium and Leaflet interact under the hood and how to use the Leaflet documentation in my usage of folium.

I guess my end goal is just to expose as much of the full functionality of Leaflet to python devs/users as possible. I've found it to be an invaluable tool in my day-to-day work fulfilling a lot of one-off map requests for my organization, and in general, being able to rapidly prototype and develop these self-contained, performant web maps that don't depend on a server implementation and new template being written is really useful.

I think being more open-ended to accept a range of inputs is a good thing, and users still have the option to refer to the Leaflet documentation and pass in those arguments directly. If it helps people work with their data more easily and lower the barrier of entry, then sounds good.

**Comment by jtbaker** [comment]
Aside - any idea what's going on with the Travis build is failing?

Errors seem to be related to a Numpy array:

- `nbconvert.preprocessors.execute.CellExecutionError: An error occurred while executing the following cell: ....
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()`

**Comment by Conengmo** [comment]
> part of me wants there to be one clear, intentional way of doing things

I agree,, and that's also in the zen of python. 

> the most clear way to do that is using the LeafletJS arguments verbatim. 

We could point to the Leaflet documentation and say to use those options for everything in folium. Problem is that would lead to horrible code: parts in Python style notation and part in Javascript style. Or we would have to convert the entire folium library to Javascript style. Just when I'm so happy that most Python projects settle on using PEP8, so there's no conflict about that in most cases.

> I guess my end goal is just to expose as much of the full functionality of Leaflet to python devs/users as possible. 

Good!

> I've found it to be an invaluable tool in my day-to-day work fulfilling a lot of one-off map requests for my organization, and in general, being able to rapidly prototype and develop these self-contained, performant web maps that don't depend on a server implementation and new template being written is really useful.

I agree. Sending people these self-contained maps works great.

> I think being more open-ended to accept a range of inputs is a good thing, and users still have the option to refer to the Leaflet documentation and pass in those arguments directly.

We should work on exposing all Leaflet options where possible. But in a sensible way, which for me means:

* no conflicting code styling in the API
* manageble bookkeeping when Leaflet updates. (We're thin on manpower so we have to keep it simple for ourselves.)

I like that we discussed some ways to do this. Let's open an issue for this in the main folium repo. We can make a plan for implementation there.
